src/main/java/org/terasology/gookeeper/system/EconomySystem.java

```diff
@@ -75,32 +93,120 @@
-                EconomyComponent economyComponent = wallet.getComponent(EconomyComponent.class);
-                if (economyComponent != null) {
-                    nuiManager.getHUD().addHUDElement("PlayerHud");
-                    setHud = true;
+    }
+
+    /**
+     * This command can be used to purchase utility blocks (such as pen blocks, visit blocks, etc.)
+     *
+     * @param client
+     * @param itemPrefabName
+     * @param amount
+     * @param shapeUriParam
+     * @return
+     */
+    @Command(shortDescription = "Purchase utility blocks",
+            requiredPermission = PermissionManager.NO_PERMISSION)
+    public String purchase(@Sender EntityRef client,
+                           @CommandParam("prefabId or blockName") String itemPrefabName,
+                           @CommandParam(value = "amount", required = false) Integer amount,
+                           @CommandParam(value = "blockShapeName", required = false) String shapeUriParam) {
+        EntityRef player = client.getComponent(ClientComponent.class).character;
+        EconomyComponent economyComponent = player.getComponent(EconomyComponent.class);
+
+        int quantityParam = amount != null ? amount : 16;
+
+        if (economyComponent.playerWalletCredit - (quantityParam * Assets.getPrefab(itemPrefabName + "Fenced").get().getComponent(PurchasableComponent.class).basePrice) > 0f) {
+            String message = blockCommands.giveBlock(client, itemPrefabName, amount, shapeUriParam);
+            if (message != null) {
+                economyComponent.playerWalletCredit -= quantityParam * Assets.getPrefab(itemPrefabName + "Fenced").get().getComponent(PurchasableComponent.class).basePrice;
+                player.saveComponent(economyComponent);
+                return "Successfully purchased " + itemPrefabName;
+            } else {
+                return "Couldn't find requested block.";
+            }
+        } else {
+            return "You dont have sufficient balance to purchase.";
+        }
+    }
+
+    /**
+     * This command can be used to purchase utility blocks (such as pen blocks, visit blocks, etc.)
+     *
+     * @param client
+     * @param itemPrefabName
+     * @param amount
+     * @param shapeUriParam
+     * @return
+     */
+    @Command(shortDescription = "Purchase utility blocks",
+            requiredPermission = PermissionManager.NO_PERMISSION)
+    public String upgrade(@Sender EntityRef client,
+                           @CommandParam("prefabId or blockName") String itemPrefabName) {
+        EntityRef player = client.getComponent(ClientComponent.class).character;
+        EconomyComponent economyComponent = player.getComponent(EconomyComponent.class);
+
+        for (int i = 0; i < inventoryManager.getNumSlots(player); i++) {
+            EntityRef itemInSlot = inventoryManager.getItemInSlot(player, i);
+            DisplayNameComponent displayNameComponent = itemInSlot.getComponent(DisplayNameComponent.class);
```

Comment:
Ah, now I see why you changed the display names. My instinct is to use `.getParentPrefab().getName()` however from memory that doesn't work on blocks when they're in the inventory.

I'm not entirely sure what to do instead so this would be another "Merge with TODO issue to fix"

Reply:
It does give an NPE currently, although this is definitely the better way of doing it